### DISCUSSION TITLE
SurfacePressureConstraint question：Can't word wthile another robot work

### DISCUSSION THREAD

**Discussion by TeoREN0217** [SofaDefrost/SoftRobots#172] [Help - Programming with SoftRobots] [answered]
Hi! 
I built a model with two ball-like SurfacePressureConstraint to imitate breath by defining an animation:
```
def myAnimation(target, factor):  
        node = rootNode
        patientNode = node.getChild('patient')
        pressureConstraint1 =patientNode.cavity1.getObject('SurfacePressureConstraint')
        pressureConstraint2 =patientNode.cavity2.getObject('SurfacePressureConstraint')
        pressureValue1 = pressureConstraint1.value.value[0]
        pressureValue2 = pressureConstraint1.value.value[0]
        pressureValue1 = 10*factor
        pressureValue2 = 10*factor
        pressureConstraint1.value = [pressureValue1]
        pressureConstraint2.value = [pressureValue2]
        print(rootNode.patient.cavity1.SurfacePressureConstraint.pressure.value)
    animate(myAnimation, {'target': rootNode}, duration=1,mode="pingpong")
```
It worked very well and it printed the value correctly. So I tried to add it to my arm-like robot scene which has 6 cableactuators and 1 slidingactuator worked in inverse mode to find a target. But after add the ball-like robot, the arm-like robot still worked but the ball-like didn't breathe. I printed the value and sometimes it's all 0 and sometimes like: 
```
0.0
0.0
2.842170943040401e-14
2.842170943040401e-14
0.0
-5.684341886080802e-14
2.842170943040401e-14
0.0
8.881784197001252e-16
-1.7763568394002505e-15
6.661338147750939e-16
0.0
1.7763568394002505e-15
2.220446049250313e-16
0.0
1.1102230246251565e-16
-2.220446049250313e-16
0.0
-4.440892098500626e-16
2.7755575615628914e-17
5.551115123125783e-17
0.0
4.440892098500626e-16
0.0
-8.881784197001252e-16
4.249463003702658e-10
-6.333760182997139e-10
2.220446049250313e-16
0.0
-5.684341886080802e-14
-2.842170943040401e-14
5.684341886080802e-14
5.684341886080802e-14
2.842170943040401e-14
-5.684341886080802e-14
0.0
0.0
2.220446049250313e-16
0.0
4.440892098500626e-16
0.0
8.881784197001252e-16
0.0
0.0
```
It's not continuouse but has some regular which make the SurfacePressureConstraint flash alternately between red and green and didn‘t belly. I'm so confused about that and I wondering  is the `GenericConstraintSolver` can't work with the `QPInverseProblemSolver`
or am I wrong about the animation method?

**Comment by EulalieCoevoet** [comment, marked as answer]
Hi @TeoREN0217,

You're right, unfortunately you can't use both `GenericConstraintSolver` and `QPInverseProblemSolver` at the same time. 
It seems that only the last constraint solver you add to your scene is taken into account. 

So what happens is that the chosen solver tries to process *all* your constraints, namely `SurfacePressureConstraint`, `PositionEffector`, `SlidingActuator`, *and* `CableActuator`, and fails. 

What you can try is use the `SurfacePressureEquality` component instead of `SurfacePressureConstraint` which will be understood by the `QPInverseProblemSolver` as a fixed constraint. Sorry there's definitively room from improvement here.

**Reply by TeoREN0217** [reply]
Thanks, I tried the inverse actuator and it worked well.
